lib/toolbar-android.js

```diff
@@ -29,53 +29,86 @@ export default function createToolbarAndroidComponent(IconNamePropType, getImage
       }
       if (props.overflowIconName) {
-        getImageSource(props.overflowIconName, size, color).then(overflowIcon => this.setState({ overflowIcon }));
+        iconsPromises.push(getImageSource(props.overflowIconName, size, color))
       }
 
-      Promise.all((props.actions || []).map(action => {
+      let actionsPromises = props.actions.map(action => {
         if (action.iconName) {
-          return getImageSource(action.iconName, action.iconSize || size, action.iconColor || color).then(icon => ({
-            ...action,
-            icon,
-          }));
+          return getImageSource(action.iconName, action.iconSize || size, action.iconColor || color)
+        } else {
+          return Promise.resolve(action);
         }
-        return Promise.resolve(action);
-      })).then(actions => this.setState({ actions }));
+      })
+
+      Promise.all([...actionsPromises, ...iconsPromises])
+        .then((iconsAndActionIcons) => {
+          let newState = {ready: true}
+
+          newState.actions = props.actions.map((action, index) => {
+            return action.iconName ? {icon: iconsAndActionIcons[index], ...action} : action;
+          })
+
+          let shift = 0
+          if (props.navIconName) {
+            newState.navIcon = iconsAndActionIcons[props.actions.length + shift++];
+          }
+          if (props.overflowIconName) {
+            newState.overflowIcon = iconsAndActionIcons[props.actions.length + shift++];
+          }
+
+          this.setState(newState)
+        });
     }
 
-    componentWillMount() {
-      this.updateIconSources(this.props);
+    constructor(props) {
+      super(props);
+      this.updateIconSources(props);
+      this.state = {
+        navIcon: undefined,
+        overflowIcon: undefined,
+        actions: undefined,
+        ready: false,
+      };
     }
 
     componentWillReceiveProps(nextProps) {
       const keys = Object.keys(IconToolbarAndroid.propTypes);
       if (!isEqual(pick(nextProps, keys), pick(this.props, keys))) {
-        let stateToEvict = {};
-        if (!nextProps.navIconName) {
-          stateToEvict.navIcon = undefined;
-        }
-        if (!nextProps.iconName) {
-          stateToEvict.icon = undefined;
-        }
-        if (this.state && Object.keys(stateToEvict).length) {
-          this.setState(stateToEvict, () => this.updateIconSources(nextProps));
-        } else {
-          this.updateIconSources(nextProps);
-        }
+        // let stateToEvict = {};
+        // if (!nextProps.navIconName) {
+        //   stateToEvict.navIcon = undefined;
+        // }
+        // if (!nextProps.overflowIconName) {
+        //   stateToEvict.overflowIcon = undefined;
+        // }
+        // if (this.state && Object.keys(stateToEvict).length) {
+        //   this.setState(stateToEvict, () => this.updateIconSources(nextProps));
+        // } else {
+        this.setState({ready: false});
```

Comment:
Hmm, not sure if this is a good idea as it will cause a flash. Isn't it better to be out of sync with props for a few milliseconds?